{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63ee3ea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Note: \"Some path to files are hardcoded. Upadte the path to run seamlessly\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e69b0fd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------Prediction of Operon for Escherichia coli K12------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cc50812b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of predicted operons for E. coli are: 780\n",
      "Predicted operons with multiple genes have been successfully saved to a text file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_ptt_file(ptt_file_path):\n",
    "    genes = []\n",
    "    with open(ptt_file_path, 'r') as file:\n",
    "        lines = file.readlines()[3:]  \n",
    "        for line in lines:\n",
    "            parts = line.strip().split('\\t')\n",
    "            start, end = map(int, parts[0].split('..'))\n",
    "            strand = parts[1]\n",
    "            genes.append((start, end, strand))\n",
    "    return genes\n",
    "\n",
    "def predict_operons(genes):\n",
    "    operons = []\n",
    "    current_operon = [genes[0]]\n",
    "\n",
    "    for i in range(1, len(genes)):\n",
    "        prev_gene = current_operon[-1]\n",
    "        current_gene = genes[i]\n",
    "\n",
    "        if current_gene[2] == prev_gene[2] and current_gene[0] - prev_gene[1] < 50:\n",
    "            current_operon.append(current_gene)\n",
    "        else:\n",
    "            if len(current_operon) > 1:  \n",
    "                operons.append(current_operon)\n",
    "            current_operon = [current_gene]\n",
    "\n",
    "    if len(current_operon) > 1:\n",
    "        operons.append(current_operon)\n",
    "\n",
    "    return operons\n",
    "\n",
    "def save_operons_to_file(operons, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for operon in operons:\n",
    "            operon_str = ', '.join([f\"{gene[0]}..{gene[1]}({gene[2]})\" for gene in operon])\n",
    "            f.write(operon_str + '\\n')\n",
    "            \n",
    "# Using functions to parse and predict operons for E. coli\n",
    "genes_E_coli = parse_ptt_file(\"E_coli_K12_MG1655.ptt/E_coli_K12_MG1655.ptt\")  # Update the path \n",
    "operons_E_coli = predict_operons(genes_E_coli)\n",
    "\n",
    "output_file_path = \"predicted_operons_E_coli.txt\"\n",
    "save_operons_to_file(operons_E_coli, output_file_path)\n",
    "\n",
    "print(f\"Total number of predicted operons for E. coli are: {len(operons_E_coli)}\")\n",
    "print(\"Predicted operons with multiple genes have been successfully saved to a text file.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cfc83050",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------Prediction of Operon for Bacillus Subtilis------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "25d36d7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of predicted operons for B_subtilis are: 771\n",
      "Predicted operons have been successfully saved to txt file\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_ptt_file(ptt_file_path):\n",
    "    genes = []\n",
    "    with open(ptt_file_path, 'r') as file:\n",
    "        lines = file.readlines()[3:]  \n",
    "        for line in lines:\n",
    "            parts = line.strip().split('\\t')\n",
    "            start, end = map(int, parts[0].split('..'))\n",
    "            strand = parts[1]\n",
    "            genes.append((start, end, strand))\n",
    "    return genes\n",
    "\n",
    "def predict_operons(genes):\n",
    "    operons = []\n",
    "    current_operon = [genes[0]]\n",
    "\n",
    "    for i in range(1, len(genes)):\n",
    "        prev_gene = current_operon[-1]\n",
    "        current_gene = genes[i]\n",
    "\n",
    "        if current_gene[2] == prev_gene[2] and current_gene[0] - prev_gene[1] < 50:\n",
    "            current_operon.append(current_gene)\n",
    "        else:\n",
    "            if len(current_operon) > 1:  \n",
    "                operons.append(current_operon)\n",
    "            current_operon = [current_gene]\n",
    "\n",
    "    if len(current_operon) > 1:\n",
    "        operons.append(current_operon)\n",
    "\n",
    "    return operons\n",
    "\n",
    "def save_operons_to_file(operons, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for operon in operons:\n",
    "            operon_str = ', '.join([f\"{gene[0]}..{gene[1]}({gene[2]})\" for gene in operon])\n",
    "            f.write(operon_str + '\\n')\n",
    "            \n",
    "ptt_file_path_b_subtilis = \"B_subtilis_168.ptt/B_subtilis_168.ptt\"  # Update the path\n",
    "\n",
    "# Using the functions\n",
    "genes_b_subtilis = parse_ptt_file(ptt_file_path_b_subtilis)\n",
    "operons_b_subtilis = predict_operons(genes_b_subtilis)\n",
    "\n",
    "# Save the predicted operons to a file\n",
    "output_file_path = \"predicted_operons_B_subtilis.txt\"\n",
    "save_operons_to_file(operons_b_subtilis, output_file_path)\n",
    "\n",
    "print(f\"Total number of predicted operons for B_subtilis are: {len(operons_b_subtilis)}\")\n",
    "print(\"Predicted operons have been successfully saved to txt file\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bed041f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------Prediction of Operon for Halobacterium------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ebe7c5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of predicted operons for Halobacterium are: 390\n",
      "Predicted operons have been successfully saved to txt file\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_ptt_file(ptt_file_path):\n",
    "    genes = []\n",
    "    with open(ptt_file_path, 'r') as file:\n",
    "        lines = file.readlines()[3:]  \n",
    "        for line in lines:\n",
    "            parts = line.strip().split('\\t')\n",
    "            start, end = map(int, parts[0].split('..'))\n",
    "            strand = parts[1]\n",
    "            genes.append((start, end, strand))\n",
    "    return genes\n",
    "\n",
    "def predict_operons(genes):\n",
    "    operons = []\n",
    "    current_operon = [genes[0]]\n",
    "\n",
    "    for i in range(1, len(genes)):\n",
    "        prev_gene = current_operon[-1]\n",
    "        current_gene = genes[i]\n",
    "\n",
    "        \n",
    "        if current_gene[2] == prev_gene[2] and current_gene[0] - prev_gene[1] < 50:\n",
    "            current_operon.append(current_gene)\n",
    "        else:\n",
    "            if len(current_operon) > 1:  \n",
    "                operons.append(current_operon)\n",
    "            current_operon = [current_gene]\n",
    "\n",
    "    if len(current_operon) > 1:\n",
    "        operons.append(current_operon)\n",
    "\n",
    "    return operons\n",
    "\n",
    "def save_operons_to_file(operons, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for operon in operons:\n",
    "            operon_str = ', '.join([f\"{gene[0]}..{gene[1]}({gene[2]})\" for gene in operon])\n",
    "            f.write(operon_str + '\\n')\n",
    "            \n",
    "# Using functions to parse and predict operons for Synechocystis\n",
    "genes_Halobacterium = parse_ptt_file(\"Halobacterium_NRC1.ptt/Halobacterium_NRC1.ptt\")  # Update the path\n",
    "operons_Halobacterium = predict_operons(genes_Halobacterium)\n",
    "\n",
    "output_file_path = \"predicted_operons_Halobacterium.txt\"\n",
    "save_operons_to_file(operons_Halobacterium, output_file_path)\n",
    "\n",
    "print(f\"Total number of predicted operons for Halobacterium are: {len(operons_Halobacterium)}\")\n",
    "print(\"Predicted operons have been successfully saved to txt file\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "47d8a522",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------Prediction of Operon for Synechocystis------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ec5e7ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of predicted operons for Synechocystis are: 484\n",
      "Predicted operons have been successfully saved to txt file\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_ptt_file(ptt_file_path):\n",
    "    genes = []\n",
    "    with open(ptt_file_path, 'r') as file:\n",
    "        lines = file.readlines()[3:]  \n",
    "        for line in lines:\n",
    "            parts = line.strip().split('\\t')\n",
    "            start, end = map(int, parts[0].split('..'))\n",
    "            strand = parts[1]\n",
    "            genes.append((start, end, strand))\n",
    "    return genes\n",
    "\n",
    "def predict_operons(genes):\n",
    "    operons = []\n",
    "    current_operon = [genes[0]]\n",
    "\n",
    "    for i in range(1, len(genes)):\n",
    "        prev_gene = current_operon[-1]\n",
    "        current_gene = genes[i]\n",
    "\n",
    "        if current_gene[2] == prev_gene[2] and current_gene[0] - prev_gene[1] < 50:\n",
    "            current_operon.append(current_gene)\n",
    "        else:\n",
    "            if len(current_operon) > 1:  \n",
    "                operons.append(current_operon)\n",
    "            current_operon = [current_gene]\n",
    "\n",
    "    if len(current_operon) > 1:\n",
    "        operons.append(current_operon)\n",
    "\n",
    "    return operons\n",
    "\n",
    "def save_operons_to_file(operons, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for operon in operons:\n",
    "            operon_str = ', '.join([f\"{gene[0]}..{gene[1]}({gene[2]})\" for gene in operon])\n",
    "            f.write(operon_str + '\\n')\n",
    "            \n",
    "# Using functions to parse and predict operons for Synechocystis\n",
    "genes_synechocystis = parse_ptt_file(\"Synechocystis_PCC6803_uid159873.ptt/Synechocystis_PCC6803_uid159873.ptt\") # Update the path\n",
    "operons_synechocystis = predict_operons(genes_synechocystis)\n",
    "\n",
    "# Save the predicted operons to a file\n",
    "output_file_path = \"predicted_operons_synechocystis.txt\"\n",
    "save_operons_to_file(operons_synechocystis, output_file_path)\n",
    "\n",
    "print(f\"Total number of predicted operons for Synechocystis are: {len(operons_synechocystis)}\")\n",
    "print(\"Predicted operons have been successfully saved to txt file\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "158f66f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------Prediction of Operon for Hoatzin (IMG id 2088090036)------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "66e78800",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of identified operons in similar contig: 753\n",
      "First few operons:\n",
      "HCP21_2940:1..159(-), HCP21_2940:192..371(-)\n",
      "HCP21_118_:42..788(+), HCP21_118_:793..2529(+), HCP21_118_:2540..3247(+)\n",
      "HCP21_6018:201..350(+), HCP21_6018:366..584(+)\n",
      "HCP21_2399:1123..1893(-), HCP21_2399:1915..2241(-)\n",
      "HCP21_2957:62..496(-), HCP21_2957:477..554(-)\n",
      "Predicted operons have been successfully saved to txt file\n"
     ]
    }
   ],
   "source": [
    "# For the Hoatzin predicting the operons based on the similar contigs that is adjacent co-directional genes \n",
    "# Having intervening distance less than 50bp distance \n",
    "\n",
    "import pandas as pd \n",
    "\n",
    "def parse_gff_for_operons(gff_file_path):\n",
    "    \n",
    "    operons = []\n",
    "    with open(gff_file_path, 'r') as file:\n",
    "        current_operon = []\n",
    "        last_end = 0\n",
    "        last_strand = ''\n",
    "        last_contig = ''\n",
    "        for line in file:\n",
    "            if line.startswith('#') or not line.strip():\n",
    "                continue  \n",
    "            parts = line.strip().split('\\t')\n",
    "            contig, feature_type, start, end, strand = parts[0], parts[2], int(parts[3]), int(parts[4]), parts[6]\n",
    "            # Only consider 'CDS' for operon prediction\n",
    "            if feature_type.lower() == 'cds':\n",
    "                # CDS co-directional with the previous and close and part of the same operon\n",
    "                if contig == last_contig and strand == last_strand and start - last_end < 50:\n",
    "                    current_operon.append((contig, start, end, strand))\n",
    "                else:\n",
    "                    if current_operon and len(current_operon) > 1:\n",
    "                        operons.append(current_operon)\n",
    "                    # Start a new operon list with the current CDS\n",
    "                    current_operon = [(contig, start, end, strand)]\n",
    "                # Update the last position and strand for the next iteration\n",
    "                last_contig, last_end, last_strand = contig, end, strand\n",
    "        # Add the final operon if it exists with more than one gene\n",
    "        if current_operon and len(current_operon) > 1:\n",
    "            operons.append(current_operon)\n",
    "    return operons\n",
    "\n",
    "def save_operons_to_file(operons, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for operon in operons:\n",
    "            # Formating each gene in the operon for output\n",
    "            operon_genes = ', '.join([f\"{contig}:{start}..{end}({strand})\" for contig, start, end, strand in operon])\n",
    "            f.write(operon_genes + '\\n\\n')\n",
    "            \n",
    "\n",
    "# Specify the path \n",
    "gff_file_path = 'Hoatzin.gff'      # Updatethe path to run seamlessly \n",
    "\n",
    "# Path to save the operon predictions\n",
    "output_file_path = 'predicted_operons_hoatzin.txt'\n",
    "\n",
    "# Parse the GFF file and predict operons\n",
    "operons = parse_gff_for_operons(gff_file_path)\n",
    "\n",
    "# Save the operons to a file\n",
    "save_operons_to_file(operons, output_file_path)\n",
    "\n",
    "# Print the total number of operons\n",
    "print(f\"Total number of identified operons in similar contig: {len(operons)}\")\n",
    "if operons:\n",
    "    print(\"First few operons:\")\n",
    "    for operon in operons[:5]:\n",
    "        print(', '.join([f\"{contig}:{start}..{end}({strand})\" for contig, start, end, strand in operon]))\n",
    "else:\n",
    "    print(\"No operons found based on the specified criteria.\")\n",
    "\n",
    "print(\"Predicted operons have been successfully saved to txt file\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
